feat(curriculum): add to-do list lab (#55886)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-stylized-to-do-list/66c051d13a6a20255a963695.md

1. Inside the unordered list, you should have 4 list items.
1. Inside each list item, there should be:
   
- An `input` element with the type `checkbox` and `id` set to a unique value.
- A `label` element with the `for` attribute set to the corresponding `input` element's `id`.
- An unordered list with the class `sub-item`.
- A list item with an `anchor` element with the class `sub-item-link` and a valid `href` within the `sub-item` unordered list.
  
1. There shouldn't be any list styles applied to the unordered lists.
2. The anchor elements shouldn't have any default text decoration.
3. You should set the text color of the links to a colour of your choice.
4. When the links are visited, the color should change to another colour of your choice.
5. When the links are hovered over, the color should change to another colour of your choice.
6. When the links are focused, there should be a coloured outline around the link.
7. When the links are clicked, the color should change to another colour of your choice.
    
# --hints--

You should have an unordered list with the class `todo-list`.

```js
assert.exists(document.querySelector('.todo-list'));
```

> 

Author: done!

You should have 4 `input` elements with the type `checkbox`.

```js
const checkboxes = document.querySelectorAll('.todo-list li input[type="checkbox"]');

assert.strictEqual(checkboxes.length, 4);
```

All of the `input` elements should have an `id`

```js
const inputs = document.querySelectorAll('input');
for (let input of inputs) {
    assert.notEmpty(input?.id);
}
```

> After the `label` elements, there should be an unordered list with the class `sub-item`.

Author: I have left this hint as it is, but made the test a bit specific:

Now, `label + ul.sub-item` would select all `ul` elements that:
- Directly follow a `label` element in the DOM.
- Have the class `sub-item`.

```js
assert.exists(document.querySelector('.todo-list li ul.sub-item li a.sub-item-link'));
```

All of the anchor elements should have a valid `href`.

```js
const links = $('a');
assert(!!links.length);
for (let link of links) {
    assert.isAbove(link.href.length, 0);
}
```

All of the `anchor` elements should have a text.

```js
const links = $('a');
assert(!!links.length);
for (let link of links) {
    assert.isAbove(link.innerText.length, 0);
}
```

All of the `label` elements should have a `for` attribute.

```js
const labels = document.querySelectorAll('label');
for (let label of labels) {
  assert.notEmpty(label?.htmlFor);
}
```

All of the `label` elements should have some text.

```js
const labels = $('label');
assert(!!labels.length);
for (let label of labels) {
    assert.isAbove(label.innerText.length, 0);
}
```

Each `a` element should have a `target` attribute with the value of `_blank`.

```js
const anchors = $('a');
assert(!!anchors.length);
for (let anchor of anchors) {
    assert.equal(anchor.target, '_blank');
}
```

The links on the page should have no underline by default.

```js

> assert.equal(new __helpers.CSSHelp(document).getStyle('.todo-list li a').getPropVal('text-decoration'), 'none' || new __helpers.CSSHelp(document).getStyle('a').getPropVal('text-decoration'), 'none');

Author: @Dario-DC 
This test didn't work.

> assert.equal(new __helpers.CSSHelp(document).getStyle('.todo-list li a').getPropVal('text-decoration'), 'none' || new __helpers.CSSHelp(document).getStyle('a').getPropVal('text-decoration'), 'none');

Author: I updated the test, but it has room for improvement.

```

The links should change color when hovered over.

```js

> assert(new __helpers.CSSHelp(document).getStyle('.sub-item-link:link')?.color || new __helpers.CSSHelp(document).getStyle('a:link')?.color);

Reviewer: Shouldn't it be `a:hover`?

---
id: 66c051d13a6a20255a963695
title: Build a Stylized To-Do list
challengeType: 14
dashedName: build-a-stylized-to-do-list
demoType: onClick
---

# --description--

In this lab, you will practice the different styles that can be applied to links when they are hovered over, focused, clicked, and visited.

> **Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. **Do not copy this demo project**.

Reviewer 2: ```suggestion
**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.
```

**User Stories:**

1. You should have an unordered list with the class `todo-list`.
2. Inside the unordered list, you should have 4 list items.
3. Inside each list item, there should be:
   
   - An `input` element with the type `checkbox` and `id` set to a unique value.
   - A `label` element with the `for` attribute set to the corresponding `input` element's `id`.
   - An unordered list with the class `sub-item`.

>    - A list item with an `anchor` element with the class `sub-item-link` and a valid `href` within the `sub-item` unordered list.

Reviewer 2: ```suggestion
   - A list item with an `anchor` element in it. The `anchor should have the class `sub-item-link`, a valid `href` value, and a `target` value that makes the link open in a new tab.
```

> 4. There shouldn't be any list styles applied to the unordered lists.
> 5. The anchor elements shouldn't have any default text decoration.
> 6. You should set the text color of the links to a colour of your choice.
> 7. When the links are visited, the color should change to another colour of your choice.
> 8. When the links are hovered over, the color should change to another colour of your choice.
> 9. When the links are focused, there should be a coloured outline around the link.
> 10. When the links are clicked, the color should change to another colour of your choice.

Reviewer 2: ```suggestion
4. You shouldn't apply any list styles to the unordered lists.
5. Your anchor elements shouldn't have any default text decoration.
6. You should set the text color of the links to a color of your choice.
7. When your links are visited, the color should change to another color of your choice.
8. When your links are hovered over, the color should change to another color of your choice.
9. When your links are focused, there should be a colored outline around the link.
10. When your links are clicked, the color should change to another color of your choice.
```

```js
const anchors = document.querySelectorAll('a');
assert(!!anchors.length);
for (let anchor of anchors) {
    assert.equal(anchor.target, '_blank');
}
```

The `a.sub-item-link` links on the page should have no underline by default.

```js
// test can be made flexible
assert(new __helpers.CSSHelp(document).getStyle('a.sub-item-link')?.textDecoration === 'none');
```

The links should change color when hovered over.

```js
assert(new __helpers.CSSHelp(document).getStyle('.sub-item-link:hover')?.color || new __helpers.CSSHelp(document).getStyle('a:hover')?.color);
```

The links should change color when they are being clicked.

```js
assert(new __helpers.CSSHelp(document).getStyle('.sub-item-link:active')?.color || new __helpers.CSSHelp(document).getStyle('a:active')?.color);
```

The links should have an outline when focused.

```js
assert(new __helpers.CSSHelp(document).getStyle('.sub-item-link:focus')?.outline || new __helpers.CSSHelp(document).getStyle('a:focus')?.outline);
```

The links should change color once visited.

```js
assert(new __helpers.CSSHelp(document).getStyle('.sub-item-link:visited')?.color || new __helpers.CSSHelp(document).getStyle('a:visited')?.color);

> ```

Reviewer 2: These tests all require using the `.sub-item-link` selector. So if I just use `a` or `a.sub-item-link`, the tests don't pass. We should either change the text to specifically say to use the `.sub-item-link` selector (might want to add it to the user stories if we go that route), or make the tests more flexible.

Author: I have made the tests flexible by targeting `a.sub-item-link`, `.sub-item-link`, and `a`.

---
id: 66c051d13a6a20255a963695
title: Build a Stylized To-Do list
challengeType: 14
dashedName: build-a-stylized-to-do-list
demoType: onClick
---

# --description--

In this lab, you will practice the different styles that can be applied to links when they are hovered over, focused, clicked, and visited.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.

**User Stories:**

1. You should have an unordered list with the class `todo-list`.

> 2. Inside the unordered list, you should have 4 list items.

Reviewer 2: ```suggestion
2. Inside the unordered list, you should have four list items.
```

3. Inside each list item, there should be:
   
   - An `input` element with the type `checkbox` and `id` set to a unique value.
   - A `label` element with the `for` attribute set to the corresponding `input` element's `id`.
   - An unordered list with the class `sub-item`.

>    - A list item with an `anchor` element in it. The `anchor should have the class `sub-item-link`, a valid `href` value, and a `target` value that makes the link open in a new tab.

Reviewer 2: ```suggestion
   - A list item with an anchor element in it. The anchor should have the class `sub-item-link`, a valid `href` value, and a `target` value that makes the link open in a new tab.
```

> 4. You shouldn't apply any list styles to the unordered lists.

Reviewer 2: I'm not sure what this one is supposed to mean. Does it mean I should use `list-style: none;` or am I not supposed to change the default list-style. If so, I would reword this to something like "You should remove the default list style from your unordered lists."

And we could add a test for it if so. Also, is it supposed to apply to all the UL's? or just the main UL? I would make it so it's just the main one. "You should remove the default list style from your `.todo-list`"

Author: This instruction is supposed to remove the default style from the main unordered list. The campers can add any style to the other list inside and it is optional.

I have updated the instruction and added a new test.
It works for both:

```css
.todo-list {
    list-style: none;
}
```

```css
.todo-list {
    list-style-type: none;
}
```
